core/templates/services/improvements.service.ts

```diff
@@ -29,6 +29,9 @@ export class ImprovementsService {
   INTERACTION_IDS_REQUIRED_TO_BE_RESOLVED = ['TextInput'];
 
   isStateForcedToResolveOutstandingUnaddressedAnswers(state: State): boolean {
+    if (state.interaction.id === null) {
+      throw new Error('Interaction Id is possibly null!');
```

Comment:
Have you investigated the callsites to see whether it's reasonable for this function to be called when the editor is loaded?

I.e. you have two choices here -- you could throw an error, or you could return false. Which is the correct thing to do, and why?

Reply:
Function is used when adding an interaction into exploration. This is an example of State we get while adding an interaction
```JSON
{
......
   "exploration": {
      "init_state_name": "Introduction",
      "param_changes": [],
      "param_specs": {},
      "states": {
         "Introduction": {
            "content": {
               "content_id": "content",
               "html": "<p>fdfdf<\/p>"
            },
            "param_changes": [],
            "interaction": {
               "id": "EndExploration",
               "customization_args": {
                  "recommendedExplorationIds": {
                     "value": []
                  }
               },
               "answer_groups": [],
               "default_outcome": null,
               "confirmed_unclassified_answers": [],
               "hints": [],
               "solution": null
            },
            "classifier_model_id": null,
            "linked_skill_id": null,
            "recorded_voiceovers": {
               "voiceovers_mapping": {
                  "content": {}
               }
            },
            "written_translations": {
               "translations_mapping": {
                  "content": {}
               }
            },
            "solicit_answer_details": false,
            "card_is_checkpoint": true,
            "next_content_id_index": 0
         }
      },
      "title": "",
      "objective": "",
      "language_code": "en",
      "correctness_feedback_enabled": false
   },
   "exploration_id": "BT64n2MjOEFa",
........
}
```
See the key `interaction` with it's child property `id` having a value `EndExploration`. So while adding an interaction a state will always have an interactionId. I don't think there would be a case when interaction id would be null. If so we would throw an error. Therfore, I think that it is a correct thing to do.

Reply:
An interaction can be null if the user hasn't specified it for a state yet. E.g. when you first open the editor, the interaction for the state is null, right?

Reply:
But at that point this function wouldn't be called right ? As the name suggests `isStateForcedToResolveOutstandingUnaddressedAnswers` it would only be called after interaction is added

Reply:
You can't type/document a function just based on what currently calls it. Other things may call it in the future, so your function should be resilient.

Also, you can't just make a guess based on the name. If I'm loading an exploration that's half-finished then it's possible that it may be called in the background when the page is loading, right (even though there is no interaction in the current state)?

So, please make the function robust.

Reply:
Ah I see, then it would be better if we would return false here if interaction id is null to avoid unnecessary errors,
 right ? Also I don't particularly see any errors at the time interaction is half loaded or when we just create an exploration.

https://user-images.githubusercontent.com/60519303/129473480-6346737c-a98b-42d6-a2bf-68c0b997ebff.mp4

Please mention any other particular usecase you'd want me to test.

Reply:
Probably yes, returning false sounds sensible.

Again, try to design so that this is error-proof in the future, not just in the present.